src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -649,6 +671,92 @@ def filter(self, col: Expression) -> "SelectStatement":
+            # There must be duplicate columns in the projection.
+            # We don't flatten when there are duplicate columns.
+            can_be_flattened = False
+            disable_next_level_flatten = True
+        elif self.flatten_disabled:
+            can_be_flattened = False
+        elif self.has_clause_using_columns and not self.snowflake_plan.session.conf.get(
+            "flatten_select_after_filter_and_orderby"
+        ):
+            # TODO: Clean up, this entire if case is parameter protection
+            can_be_flattened = False
+        elif self.where and (
+            (subquery_dependent_columns := derive_dependent_columns(self.where))
+            in (COLUMN_DEPENDENCY_DOLLAR, COLUMN_DEPENDENCY_ALL)
+            or any(
+                new_column_states[_col].change_state == ColumnChangeState.NEW
+                for _col in subquery_dependent_columns.intersection(
+                    new_column_states.active_columns
+                )
+            )
+        ):
+            can_be_flattened = False
+        elif self.order_by and (
+            (subquery_dependent_columns := derive_dependent_columns(*self.order_by))
+            in (COLUMN_DEPENDENCY_DOLLAR, COLUMN_DEPENDENCY_ALL)
+            or any(
+                new_column_states[_col].change_state
+                in (ColumnChangeState.CHANGED_EXP, ColumnChangeState.NEW)
+                for _col in subquery_dependent_columns.intersection(
+                    new_column_states.active_columns
+                )
+            )
+        ):
+            can_be_flattened = False
+        else:
+            can_be_flattened = can_select_statement_be_flattened(
+                self.column_states, new_column_states
+            )
+
+        if can_be_flattened:
+            final_projection = []
+
+            for col, state in new_column_states.items():
+                if state.change_state in (
+                    ColumnChangeState.CHANGED_EXP,
+                    ColumnChangeState.NEW,
+                ):
+                    final_projection.append(copy(state.expression))
+                elif state.change_state == ColumnChangeState.UNCHANGED_EXP:
+                    final_projection.append(
+                        copy(self.column_states[col].expression)
+                    )  # add subquery's expression for this column name
+
+            new = copy(self)
+            new.from_ = self.from_.to_subqueryable()
+            new.pre_actions = new.from_.pre_actions
+            new.post_actions = new.from_.post_actions
+            new.projection = final_projection
+            if self.exclude is None:
+                new.exclude = cols
+            else:
+                new.exclude = self.exclude + cols
+
+        else:
+            new = SelectStatement(
+                from_=self.to_subqueryable(),
+                projection=kept_cols,
+                exclude=cols,
```

Comment:
Will we eventually do some magic between these two such that we get the correct projection columns?

Reply:
The exclude parameter is primarily used to generate the correct SQL statement, does not play a role in other computation (projection does). As long as the projection parameter is accurate (contains columns that should remain in the dataframe), we will get the correct projections.

Reply:
Let's say we eventually have `projection` that is `[a, b, c]`, and `exclude` is `[a]`. Do we squash the result such that we know the result is `[b, c]` (because we try to find the intersection), or is this not a legit input?

Reply:
Projection would be `[b, c]` if exclude is `[a]`, because of the way projection and exclude are computed. In the `drop()` function in `dataframe.py`, `kept_col_names` and `remove_col_names` are mutually exclusive (and these eventually form the variables `projection` and `exclude`)